Mathlib/MeasureTheory/Measure/Tight.lean

```diff
@@ -0,0 +1,483 @@
+/-
+Copyright (c) 2024 Josha Dekker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Josha Dekker
+-/
+import Mathlib.MeasureTheory.Measure.Regular
+import Mathlib.Topology.Metrizable.Uniformity
+import Mathlib.Analysis.SpecificLimits.Basic
+
+/-!
+# (Pre-)tight measures
+The key definition of interest here is that of tight measures, `IsTight`. We first introduce two
+weaker notions, `IsSeparablySupported` and `IsPretight`, which are equivalent on complete metric
+spaces. We provide some basic API for these notions and prove Ulam's tightness theorem
+(`of_isSeparableSpace_completeSpace`) and its strengthening `of_isSeparablySupported_completeSpace`.
+
+## Main definitions
+* `MeasureTheory.Measure.IsSeparablySupported`: A measure `μ` is separably supported if there is a
+  separable set `S` such that `μ Sᶜ = 0`.
+* `MeasureTheory.Measure.IsPretight`: A measure `μ` is pre-tight if for all `0 < ε`, there exists
+  `K` totally bounded such that `μ Kᶜ ≤ ε`.
+* `MeasureTheory.Measure.IsTight`: A measure `μ` is tight if for all `0 < ε`, there exists `K`
+  compact such that `μ Kᶜ ≤ ε`. This is defined in terms of filters.
+  `iff_exists_isCompact_measure_compl_le` establishes equivalence with the usual definition.
+
+## Main statements
+
+* `of_isSeparableSpace_completeSpace`: Ulam's tightness theorem: a finite measure on a complete
+  separable metric space is tight.
+* `of_isSeparablySupported_completeSpace`: A strengthening of Ulam's tightness theorem: a finite,
+  separably supported measure on a complete metric space is tight.
+* `of_innerRegular`: Every finite, inner-regular measure on a T2 space is tight.
+
+## Notation
+
+
+## Implementation details
+
+
+-/
+open Topology Filter Uniformity Uniform MeasureTheory Set
+open scoped ENNReal
+
+namespace MeasureTheory.Measure
+
+variable {α ι : Type*} [MeasurableSpace α] {μ : Measure α}
+
+/-- For a finite measure `μ`, we can extract from a countable cover that has full measure, a finite
+cover of accumulated sets that has `ε`-almost full measure. -/
+lemma exists_sub_le_measure_accumulate [IsFiniteMeasure μ] (K : ℕ → Set α)
+    (h : μ (⋃ n, K n) = μ Set.univ) {ε : ENNReal} (hε : ε > 0) :
+    ∃ n, μ Set.univ - ε ≤ μ (Set.Accumulate K n) := by
```

Comment:
Instead of writing this as "there exists some `n` such that", better spell it as "for all large enough `n`, then ..." (written in Lean as ∀ᶠ n in atTop, ...). When you need to access such an `n`, use `(exists_sub_le_measure_accumulate ...).exists`. Same thing in the next two lemmas.